### ISSUE TITLE
Using ensure_future() not supported on Ubuntu 14.04

### ISSUE THREAD

**Issue by n1koo** [mbevand/silentarmy#8]
This might be design, but usage of `ensure_future()` (eg. https://github.com/mbevand/silentarmy/blob/master/silentarmy#L105 ) in `silentarmy` requires Python 3.4.4+ .

Trusty only has up to 3.4.3 available (http://packages.ubuntu.com/trusty-updates/python3.4)

Maybe deprecate support for 14.04 ?

```
n1ko@orangeminer:~/silentarmy$ ./silentarmy
Traceback (most recent call last):
  File "./silentarmy", line 549, in <module>
    main()
  File "./silentarmy", line 546, in main
    Silentarmy(opts).run()
  File "./silentarmy", line 256, in run
    asyncio.ensure_future(self.reconnect())
AttributeError: 'module' object has no attribute 'ensure_future'
```
```
n1ko@orangeminer:~/silentarmy$ python3 --version
Python 3.4.3
```

**Comment by mbevand** [comment]
Oh what a bummer! I wanted to support 14.04 and did not think my Python would be incompatible. Oh well I documented it in the readme. I certainly want to try to make 14.04 fully supported.

**Comment by n1koo** [comment]
Seems to work in 14.04 after installing python3.5 manually. Seeing https://github.com/mbevand/silentarmy/issues/9 but mining is totally fine and getting +40% boost compared to `zogminer` so this is pretty huge!

**Comment by n1koo** [comment]
For anyone stumbling upon here looking at the same issue, heres a gist to install deps not available in 14.04 and also edit the `silentarmy` to use the custom 3.5 rather than 3.4 from system: https://gist.github.com/n1koo/7c0455b674d3c3a9d4cccf53837b270e

**Comment by mbevand** [comment]
@n1koo Thank you for documenting the workaround in so much details! Also I added a pointer to this github issue: https://github.com/mbevand/silentarmy/commit/6a038fe6650ac0256e6a360cdac4baece1e1856c

**Comment by lhl** [comment]
Probably should have looked at the issues list earlier, but just in case it helps anyone before mrb gets a chance to review, I have a pending PR: #27 that should get the miner working in 3.3-3.4.3; I've tested on 3.3.6 in Arch but don't have any Ubuntu 14.04 LTS systems so maybe someone can give it a spin overnight and comment in the PR issue if it doesn't work. You can build/compile from https://github.com/lhl/silentarmy which is an up-to-date fork as of right now except for the asyncio backport code.

**Comment by mbevand** [comment]
Yes I am going to look into lhl's pull request tomorrow.

**Comment by PabloCastellano** [comment]
I confirm that @lhl's branch works fine. I had to revert the sys.path comit and manually copy sa-solver to lib/asyncio though.

**Comment by lhl** [comment]
Hmm @PabloCastellano, that's interesting - so you reverted https://github.com/mbevand/silentarmy/pull/27/commits/0863a831386ab0f05410a061648a114fb73b4ab1?

The reason that I had to swap to `sys.path.append()` instead of `sys.path.insert()` is because [silentarmy depends on `sys.path[0]` to assign the `sa-solver`location](https://github.com/mbevand/silentarmy/blob/master/silentarmy#L133
). By default `sys.path[0]` is usually ''.  Arguably, `sa-solver` PATH should probably be assigned by setting a HOME or SOLVER path like how I set the `LIBPATH` with `os.path.dirname(os.path.realpath(__file__))`, but I was doing my best to design this patch to be the least invasive possible solution.
